onadata/apps/api/viewsets/xform_viewset.py

```diff
@@ -749,6 +763,74 @@ def destroy(self, request, *args, **kwargs):
 
         return Response(status=status.HTTP_204_NO_CONTENT)
 
+    @action(methods=['GET'], detail=True)
+    def versions(self, request, *args, **kwargs):
+        xform = self.get_object()
+        versions_cache_key = f'{XFORM_DATA_VERSIONS}{xform.pk}'
+        version_id = kwargs.get('version_id')
+        job_uuid = request.query_params.get('job_uuid')
+
+        versions = cache.get(versions_cache_key, [])
+
+        # Retrieve a list of all available versions from the XForms'
+        # instances if the number of submissions are not more than the
+        # retrieval threshold
+        if not xform.num_of_submissions > getattr(
+            settings, "SUBMISSION_RETRIEVAL_THRESHOLD", 10000
+        ) and not versions:
+            versions = list(
+                xform.instances.filter(
+                    deleted_at__isnull=True
+                ).values('version').annotate(total=Count('version'))
+            )
+            cache.set(versions_cache_key, versions)
+
+        if versions:
+            versions = [version.get('version') for version in versions]
+
+        stored_versions = XFormVersion.objects.filter(
+            xform=xform
+        ).values_list('version')
+        stored_versions = [version[0] for version in versions]
+        stored_versions.append(xform.version)
+
+        for version in stored_versions:
+            if version not in versions:
+                versions.append(version)
+
+        if not version_id:
+            return Response(data=versions, status=status.HTTP_200_OK)
+
+        if versions and version_id not in versions:
+            return Response(status=status.HTTP_400_BAD_REQUEST)
+        elif job_uuid:
+            resp = tasks.get_async_status(job_uuid)
+            return Response(data=resp, status=status.HTTP_202_ACCEPTED)
+        else:
+            if version_id == xform.version:
+                data = {
+                    'xml': xform.xml
+                }
```

Comment:
I see the need to get the XForm XML when you hit `/api/v1/forms/[xform pk]/versions/[version pk].xml. Why would we want the XML as part of a JSON response?